src/yunohost/data_migrations/0021_migrate_to_bullseye.py

```diff
@@ -30,6 +30,59 @@
 N_NEXT_DEBAN = 11
 N_NEXT_YUNOHOST = 11
 
+VENV_BACKUP_SUFFIX = "_BACKUP_VENV"
+VENV_REQUIREMENTS_SUFFIX = "_req.txt"
+VENV_IGNORE = "ynh_migration_no_regen"
+
+def _get_all_venvs(dir,level=0,maxlevel=2):
+    result = []
+    for file in os.listdir(dir):
+        path = os.path.join(dir,file)
+        if os.path.isdir(path):
+            if os.path.isfile(os.path.join(path,VENV_IGNORE)):
+                continue
+            activatepath = os.path.join(path,"bin","activate")
+            if os.path.isfile(activatepath):
+                content = read_file(activatepath)
+                if "VIRTUAL_ENV" and "PYTHONHOME" in content:
+                    result.append(path)
+                    continue
+            if level<maxlevel:
+                result+=_get_all_venvs(path,level=level+1)
+    return result
+
+def _generate_requirements():
+
+    venvs = _get_all_venvs("/opt/")+_get_all_venvs("/var/www/")
+    for venv in venvs:
+        # Generate a requirements file from venv
+        os.system(f"bash -c 'source {venv}/bin/activate && pip freeze > {venv}{VENV_REQUIREMENTS_SUFFIX} && deactivate'")
```

Comment:
I am not totally sure, but i think this is better:
```suggestion
        os.system(f"{venv}/bin/pip freeze > {venv}{VENV_REQUIREMENTS_SUFFIX}")
```

Missing try except clause ?

Reply:
What if we have no space left on device ?

Reply:
To be sure we keep the unchange version